examples/barotropicqg_betaforced.jl

```diff
@@ -69,7 +68,7 @@ seed!(1234) # reset of the random number generator for reproducibility
 nothing # hide
 
 # Next we construct function `calcF!` that computes a forcing realization every timestep
-function calcFq!(Fh, sol, t, cl, v, p, g)
+function calcFq!(Fh, sol, t, cl, vs, pr, gr)
   ξ = ArrayType(dev)(exp.(2π*im*rand(Float64, size(sol)))/sqrt(cl.dt))
```

Comment:
Does this allocate a new array every time-step?

You may want to allocate this array outside of the function, and then reference it as a global. I think declaring it `const` will ensure that it runs on the GPU.

Another option is to make `calcFq` a callable object that carries around a reference to `ξ` (rather than referencing it as a global). The complication with this approach is that I think you will need to use `Adapt` to get it to work on the GPU.

It's probably important to introduce what we think is best practice for this problem in the example. If its complicated, I think that means we simply need to add more simple examples to bridge the gap between simple cases and this sophisticated case.

Reply:
I guess referencing `ξ` as a global is probably a simple and adequate solution for this problem. 

I'm worried this forcing function won't work (at least not generically, or for large problems) on the GPU if it allocates every time-step, as there is no garbage collection on the GPU.

Also, the syntax on this fairly complex line is very compact, which makes it difficult to read. Putting some white space between the arithmetic operations will make this part of the code easier to read and understand.

Reply:
Nice catch!

Reply:
One solution is to avoid `ξ` whatsoever by:
```julia
  @. Fh = ArrayType(dev)(exp.(2π*im*rand(Float64, size(sol)))/sqrt(clock.dt))*sqrt(forcing_spectrum)
```

But that seems cumbersome, right?

Reply:
hm... I tried `@btime` and it made no difference...

Reply:
Your solution still allocates memory. Memory is allocated by this snippet:

`rand(eltype(gr), size(sol))`

Which is embedded within the expression above. This function creates an array of `size(sol)`.

This function always executes on the CPU, as well.

Note that (the last time I checked) it is not possible to generate random numbers on the GPU within a kernel (and therefore without allocation GPU memory). This function will be slow on the GPU. It may be best to write this code only to execute on the CPU, so that people are not misled into thinking it would be a good idea to run this on the GPU. I'm pretty sure this code would be very slow on the GPU.

One way to avoid allocating memory on the CPU is probably to write a function that executes a loop over the two dimensions of `ξ`. For example, you can fill an array with random values by writing

```julia
for i in eachindex(a)
    @inbounds a[i] = rand()
end
```

Reply:
It's also important to note that `ArrayType(GPU())(a)` creates a new array that has the data in `a` on the GPU, and therefore allocates memory on the GPU. Because there is no garbage collection on the GPU, the GPU will run out of memory when this function is executed many times.

Reply:
This also works:

```julia
julia> @time a .= rand(Float64, size(a));
  0.000447 seconds (9 allocations: 512.297 KiB)
```

Reply:
Here's an example:

```julia
julia> @time a .= rand(Float64, size(a));
  0.000175 seconds (9 allocations: 512.297 KiB)
```

```julia
julia> function fillrand!(a)
       for i in eachindex(a)
       @inbounds a[i] = rand()
       end
       end

julia> @time fillrand!(a)
  0.009296 seconds (6.76 k allocations: 315.136 KiB)

julia> @time fillrand!(a)
  0.000686 seconds (4 allocations: 160 bytes)
```

The function that allocates memory is actually faster in this case... but it might slow down time-stepping. I'm not sure.

Reply:
Ah, best of all:

```julia
using Random

a = zeros(256, 256)

@time rand!(a)

@time rand!(a)
```

I get 

```julia
julia> @time rand!(a);
  0.000087 seconds (4 allocations: 160 bytes)
```